src/auth/session.rs

```diff
@@ -0,0 +1,80 @@
+use crate::auth::AuthenticationError;
+use crate::database::models::session_item::{Session, SessionBuilder};
+use crate::database::models::UserId;
+use crate::util::env::parse_var;
+use actix_web::HttpRequest;
+use rand::distributions::Alphanumeric;
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use woothee::parser::Parser;
+
+pub async fn issue_session(
+    req: HttpRequest,
+    user_id: UserId,
+    transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    redis: &deadpool_redis::Pool,
+) -> Result<Session, AuthenticationError> {
+    let conn_info = req.connection_info();
+    let ip_addr = if parse_var("CLOUDFLARE_INTEGRATION").unwrap_or(false) {
+        if let Some(header) = req.headers().get("CF-Connecting-IP") {
+            header.to_str().ok()
+        } else {
+            conn_info.peer_addr()
+        }
+    } else {
+        conn_info.peer_addr()
+    };
+
+    let country = req
+        .headers()
+        .get("cf-ipcountry")
+        .and_then(|x| x.to_str().ok());
+    let city = req.headers().get("cf-ipcity").and_then(|x| x.to_str().ok());
+
+    let user_agent = req
+        .headers()
+        .get("user-agent")
+        .and_then(|x| x.to_str().ok())
+        .ok_or_else(|| AuthenticationError::InvalidCredentials)?;
```

Comment:
Could we make this a separate err? If I were trying to log in and I forgot the user agent, invalid credentials being the error would be confusing and make me think I had the wrong password or mistyped my token or smthn.

If we dont anticipate this being used by the API ever directly then its fine

Reply:
it won't be in the API, so no need